nilearn/regions/connectivity.py

```diff
@@ -0,0 +1,392 @@
+"""Methods For defining voxelwise connectivity"""
+# Author: Derek Pisner
+
+import numpy as np
+from scipy import divide, prod
+from .._utils.niimg import _safe_get_data
+
+def indx_1dto3d(idx, sz):
+    """
+    Translate 1D vector coordinates to 3D matrix coordinates for a 3D matrix
+    of size sz.
+
+    Parameters
+    ----------
+    idx : array
+        A 1D numpy coordinate vector.
+    sz : array
+        Shape of 3D matrix idx.
+
+    Returns
+    -------
+    x : int
+        x-coordinate of 3D matrix coordinates.
+    y : int
+        y-coordinate of 3D matrix coordinates.
+    z : int
+        z-coordinate of 3D matrix coordinates.
+    """
+
+    x = divide(idx, prod(sz[1:3]))
+    y = divide(idx - x * prod(sz[1:3]), sz[2])
+    z = idx - x * prod(sz[1:3]) - y * sz[2]
+    return x, y, z
+
+
+def indx_3dto1d(idx, sz):
+    """
+    Translate 3D matrix coordinates to 1D vector coordinates for a 3D matrix
+    of size sz.
+
+    Parameters
+    ----------
+    idx : array
+        A 3D numpy array of matrix coordinates.
+    sz : array
+        Shape of 3D matrix idx.
+
+    Returns
+    -------
+    idx1 : array
+        A 1D numpy coordinate vector.
+    """
+
+    if np.linalg.matrix_rank(idx) == 1:
+        idx1 = idx[0] * prod(sz[1:3]) + idx[1] * sz[2] + idx[2]
+    else:
+        idx1 = idx[:, 0] * prod(sz[1:3]) + idx[:, 1] * sz[2] + idx[:, 2]
+    return idx1
+
+
+def make_local_connectivity_scorr(func_img, clust_mask_img, thresh):
+    """
+    Constructs a spatially constrained connectivity matrix from a fMRI dataset.
+    The weights w_ij of the connectivity matrix W correspond to the
+    spatial correlation between the whole brain FC maps generated from the
+    time series from voxel i and voxel j. Connectivity is only calculated
+    between a voxel and the 27 voxels in its 3D neighborhood
+    (face touching and edge touching).
+
+    Parameters
+    ----------
+    func_img : Nifti1Image
+        4D Nifti1Image containing fMRI data.
+    clust_mask_img : Nifti1Image
+        3D NIFTI file containing a mask, which restricts the voxels used in
+        the analysis.
+    thresh : str
+        Threshold value, correlation coefficients lower than this value
+        will be removed from the matrix (set to zero).
+
+    Returns
+    -------
+    W : Compressed Sparse Matrix
+        A Scipy sparse matrix, with weights corresponding to the spatial
+        correlation between the time series from voxel i and voxel j
+
+    References
+    ----------
+    .. [1] Craddock, R. C., James, G. A., Holtzheimer, P. E., Hu, X. P., &
+      Mayberg, H. S. (2012). A whole brain fMRI atlas generated via
+      spatially constrained spectral clustering. Human Brain Mapping.
+      https://doi.org/10.1002/hbm.21333
+
+    """
+    from scipy.sparse import csc_matrix
+    from itertools import product
+
+    neighbors = np.array(
```

Comment:
explain what this is doing

Reply:
Cleanly creates the 3d neighborhood described above, rather than explicitly writing out the array in the function:
array([[-1, -1, -1],
       [ 0, -1, -1],
       [ 1, -1, -1],
       [-1,  0, -1],
       [ 0,  0, -1],
       [ 1,  0, -1],
       [-1,  1, -1],
       [ 0,  1, -1],
       [ 1,  1, -1],
       [-1, -1,  0],
       [ 0, -1,  0],
       [ 1, -1,  0],
       [-1,  0,  0],
       [ 0,  0,  0],
       [ 1,  0,  0],
       [-1,  1,  0],
       [ 0,  1,  0],
       [ 1,  1,  0],
       [-1, -1,  1],
       [ 0, -1,  1],
       [ 1, -1,  1],
       [-1,  0,  1],
       [ 0,  0,  1],
       [ 1,  0,  1],
       [-1,  1,  1],
       [ 0,  1,  1],
       [ 1,  1,  1]])